proposals/ideas/xxx-flashflow.txt

```diff
@@ -0,0 +1,713 @@
+background traffic bytes sent in the last second (since the last MSM_BG cell).
+recv_bg_bytes is the same but for received bytes.
+
+The payload of MSM_ERR cells:
+
+  - err_code [1 byte]
+  - err_str  [possibly zero-len null-terminated string]
+
+The error code is one of:
+
+  [... XXX TODO ...]
+  255 -- OTHER
+
+The error string is optional in all cases. It isn't present if the first byte
+of err_str is null, otherwise it is present. It ends at the first null byte or
+the end of the cell, whichever comes first.
+
+3.1.4 Measurement Mode
+
+The relay considers the measurement to have started the moment it receives the
+first MSM_ECHO cell from any measurer. At this point, the relay
+
+  - Starts a repeating 1s timer on which it will report the amount of
+    background traffic to the coordinator over the coordinator's connection.
+  - Enters "measurement mode" and limits the amount of background traffic it
+    handles according to the torrc option/consensus parameter.
+
+The relay decrypts and echos back all MSM_ECHO cells it receives on
+measurement connections until it has reported its amount of background traffic
+the same number of times as there are seconds in the measurement (e.g. 30
+per-second reports for a 30 second measurement). After sending the last
+MSM_BG cell, the relay drops all buffered MSM_ECHO cells, closes all
+measurement connections, and exits measurement mode.
+
+During the measurement the relay targets a ratio of background traffic to
+measurement traffic as specified by a consensus parameter/torrc option. For a
+given ratio r, if the relay has handled x cells of measurement traffic recently,
+Tor then limits itself to y = xr/(1-r) cells of non-measurement traffic this
+scheduling round. The target will enforce that a minimum of 10 Mbit/s of
+measurement traffic is recorded since the last background traffic scheduling
+round to ensure it always allows some minimum amount of background traffic.
+
+3.2 FlashFlow Components
+
+The FF coordinator and measurer code will reside in a FlashFlow repository
+separate from little-t tor.
+
+There are three notable parameters for which a FF deployment must choose
+values. They are:
+
+  - The number of sockets, s, the measurers should open, in aggregate, with
+    the target relay. We suggest s=160 based on the FF paper.
+  - The bandwidth multiplier, m. Given an existing capacity estimate for a
+    relay, z, the coordinator will instruct the measurers to, in aggregate,
+    send m*z Mbit/s to the target relay. We recommend m=2.25.
+  - The measurement duration, d. Based on the FF paper, we recommend d=30
+    seconds.
+
+The rest of this section first discusses notable functions of the FlashFlow
+coordinator, then goes on to discuss FF measurer code that will require
+supporting tor code.
+
+3.2.1 FlashFlow Coordinator
+
+The coordinator is responsible for scheduling measurements, aggregating
+results, and producing v3bw files.  It needs continuous access to new
```

Comment:
Maybe you should mention how the coordinator aggregates results (i.e. using the median per-second total throughput). Similarly, maybe you should mention that measurers send their per-second totals to the coordinator at the end of the measurement (or is it during the measurement?).